### ISSUE TITLE
Describe how to debug framework with deepkit-framework repo

### ISSUE THREAD

**Issue by colorcube** [deepkit/deepkit-framework#399]
- I'm not familar with lerna
- I've read DEVELOPMENT.md
- installation worked

`npx lerna run build `
seems to build most packages
`npm run tsc-watch`
seem to work

I would expect that I can start the example-app to test things out. 
But that doesn't work and don't seems to be intended.

A short description how to test out features for debugging would be nice.

**Comment by marcj** [comment]
> But that doesn't work and don't seems to be intended.

What error did you get?

**Comment by colorcube** [comment]
in one shell I start tsc-watch in the second I try to run the app

```shell
> npm run tsc-watch
Starting compilation in watch mode...
```

```shell
 /var/www/deepkit-framework/packages/example-app ‹master*› » npm run app

> @deepkit/example-server@1.0.1-alpha.0 app
> node -r ts-node/register --preserve-symlinks  app.ts

Error: Cannot find module '/var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/dist/cjs/index.js'
    at createEsmNotFoundErr (node:internal/modules/cjs/loader:960:15)
    at finalizeEsmResolution (node:internal/modules/cjs/loader:953:15)
    at resolveExports (node:internal/modules/cjs/loader:482:14)
    at Function.Module._findPath (node:internal/modules/cjs/loader:522:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:919:27)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/var/www/deepkit-framework/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/var/www/deepkit-framework/packages/example-app/app.ts:2:1) {
  code: 'MODULE_NOT_FOUND',
  path: '/var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/package.json'
}
```

**Comment by colorcube** [comment]
to import from example-app/node_modules/ makes no sense to me

**Comment by marcj** [comment]
do `/var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/dist/cjs/index.js` 
or `/var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework`exist?

**Comment by marcj** [comment]
also does `/var/www/deepkit-framework/packages/framework/dist/cjs/index.js` exist?

**Comment by marcj** [comment]
side note (unlikely to be related to this issue): please don't use `npx`. It might use a Lerna version we do not support. We have `npm run build` for that which uses the correct Lerna version.

**Comment by colorcube** [comment]
no cjs dir but an almost empty esm dir

```shell
» ls /var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/dist/
esm
» ls /var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/dist/esm
package.json
```

**Comment by marcj** [comment]
so that means that `packages/framework` hasn't been built. `npm run tsc-watch` builds all packages and creates these files (e.g. `packages/framework/dist/cjs/index.js`), so you have to wait until the `tsc-watch` command prints that everything went well. this takes several minutes for the very first run

**Comment by colorcube** [comment]
tsc-watch in root dir?

```
/var/www/deepkit-framework ‹master*› » npm run tsc-watch
[22:24:06] Starting compilation in watch mode...

[22:24:06] Found 0 errors. Watching for file changes.
```

This in root dir
```
/var/www/deepkit-framework ‹master*› » npm run build
```

builds most packages
but failed here:

```

Build at: 2023-01-05T21:20:44.495Z - Hash: 62f8666ca47923a8 - Time: 113886ms
lerna ERR! npm run build exited 2 in '@deepkit/example-server'
lerna ERR! npm run build stdout:

> @deepkit/example-server@1.0.1-alpha.0 build
> tsc

node_modules/@deepkit/api-console-module/node_modules/@deepkit/template/src/elements.d.ts(15,10): error TS2300: Duplicate identifier 'IfEquals'.
node_modules/@deepkit/api-console-module/node_modules/@deepkit/template/src/elements.d.ts(19,10): error TS2300: Duplicate identifier 'WritableKeysOf'.
node_modules/@deepkit/api-console-module/node_modules/@deepkit/template/src/elements.d.ts(23,10): error TS2300: Duplicate identifier 'NoFunctions'.
node_modules/@deepkit/api-console-module/node_modules/@deepkit/template/src/elements.d.ts(25,10): error TS2300: Duplicate identifier 'ExtractProps'.
node_modules/@deepkit/template/src/elements.d.ts(15,10): error TS2300: Duplicate identifier 'IfEquals'.
node_modules/@deepkit/template/src/elements.d.ts(19,10): error TS2300: Duplicate identifier 'WritableKeysOf'.
node_modules/@deepkit/template/src/elements.d.ts(23,10): error TS2300: Duplicate identifier 'NoFunctions'.
node_modules/@deepkit/template/src/elements.d.ts(25,10): error TS2300: Duplicate identifier 'ExtractProps'.
slim.ts(29,37): error TS2419: Types of construct signatures are incompatible.
  Type 'new (req: IncomingMessage) => HttpResponse' is not assignable to type 'new <Request extends IncomingMessage = IncomingMessage>(req: Request) => ServerResponse<Request>'.
    Construct signature return types 'HttpResponse' and 'ServerResponse<Request>' are incompatible.
      The types of 'req' are incompatible between these types.
        Type 'IncomingMessage' is not assignable to type 'Request'.
          'IncomingMessage' is assignable to the constraint of type 'Request', but 'Request' could be instantiated with a different subtype of constraint 'IncomingMessage'.
slim.ts(31,54): error TS2352: Conversion of type 'ServerResponse<IncomingMessage> & { req: HttpRequest; }' to type 'HttpResponse' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property 'status' is missing in type 'ServerResponse<IncomingMessage> & { req: HttpRequest; }' but required in type 'HttpResponse'.
lerna ERR! npm run build exited 2 in '@deepkit/example-server'
lerna WARN complete Waiting for 1 child process to exit. CTRL-C to exit immediately.
amun :: /var/www/deepkit-framework ‹master*› »  
```

**Comment by colorcube** [comment]
```
deepkit-framework/packages/example-app ‹master*› » npm run app                                                                                                                      

> @deepkit/example-server@1.0.1-alpha.0 app
> node -r ts-node/register --preserve-symlinks  app.ts

Error: Cannot find module '/var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/framework/node_modules/@deepkit/api-console-module/dist/cjs/index.js'
```
The path is too nested?!

Here are the builds:

```
deepkit-framework/packages/example-app ‹master*› » ls /var/www/deepkit-framework/packages/example-app/node_modules/@deepkit 
api-console-module  app  broker  bson  core  core-rxjs  event  framework  http  injector  logger  mongo  orm  rpc  rpc-tcp  sql  sqlite  stopwatch  template  type  workflow

 deepkit-framework/packages/example-app ‹master*› » ls /var/www/deepkit-framework/packages/example-app/node_modules/@deepkit/api-console-module 
index.ts  node_modules  package.json  package-lock.json  README.md  src  tests  tsconfig.esm.json  tsconfig.json
```

**Comment by marcj** [comment]
>The path is too nested?!

Paths are fine, they are exactly what they should be. Anything else (like importing modules relative via `../../api-console-module/dist/cjs/index.js`) will lead to broken module resolution as it will lose the origin context.

I've found the issue, fixed in https://github.com/deepkit/deepkit-framework/commit/dbb12f4b237ea4e4d64b00945af6a02a2e0c0519

**Comment by marcj** [comment]
You have to re-run the steps in development.md, specifically the `npm run link` in this case

**Comment by colorcube** [comment]
did it. Now one step further

```
lerna ERR! npm run build exited 2 in '@deepkit/example-server'
lerna ERR! npm run build stdout:

> @deepkit/example-server@1.0.1-alpha.0 build
> tsc

slim.ts(29,37): error TS2419: Types of construct signatures are incompatible.
  Type 'new (req: IncomingMessage) => HttpResponse' is not assignable to type 'new <Request extends IncomingMessage = IncomingMessage>(req: Request) => ServerResponse<Request>'.
    Construct signature return types 'HttpResponse' and 'ServerResponse<Request>' are incompatible.
      The types of 'req' are incompatible between these types.
        Type 'IncomingMessage' is not assignable to type 'Request'.
          'IncomingMessage' is assignable to the constraint of type 'Request', but 'Request' could be instantiated with a different subtype of constraint 'IncomingMessage'.
slim.ts(31,54): error TS2352: Conversion of type 'ServerResponse<IncomingMessage> & { req: HttpRequest; }' to type 'HttpResponse' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property 'status' is missing in type 'ServerResponse<IncomingMessage> & { req: HttpRequest; }' but required in type 'HttpResponse'.
```

**Comment by colorcube** [comment]
without slim.ts the build works
and the example-app too

**Comment by marcj** [comment]
Yeah, `slim.ts` is currently buggy with newest node types as the type of IncomingMessage has changed. not sure right now what a fix for slim.ts could be, so it's probably best to ignore it or throw around some `as any`.

**Comment by colorcube** [comment]
Thanks for the help!

I try to provide a few additional lines for development.md which I would understand ;-) if you don't mind

**Comment by colorcube** [comment]
- I have Python3 and it works - Python2 seems not needed?
- I don't know when and why `tsc-watch:esm` is needed

My suggestion:

# Development

## Prerequisites
Deepkit uses NPM and Lerna to manage this monorepo. Local package linking is managed through the [npm-local-development](https://www.npmjs.com/package/npm-local-development) CLI.

Make sure `libpq5` and `libpq-dev` are installed, and `python` refers Python2.

Node v18 is needed.

## Getting Started

```shell
git clone https://github.com/deepkit/deepkit-framework.git
cd deepkit-framework
npm install
npm run bootstrap
npm run link
npm run install-compiler
```

When installation is finished you can build the packages:

```shell
deepkit-framework » npm run build
```

This could take several minutes. 
You should see the build messages and a _success_ summary in the end:

```shell
lerna success run Ran npm script 'build' in 36 packages in 83.9s:
lerna success - @deepkit/angular-universal
lerna success - @deepkit/api-console-api
...
```

If everything went fine you can try out the example app:

```shell
deepkit-framework » cd packages/example-app
deepkit-framework/packages/example-app » npm run app
```

That should give you a _usage_ message of the app. 

To start the app server:

```shell
deepkit-framework/packages/example-app » npm run start
```

```shell
...
2023-01-05T23:22:02.199Z [LOG] HTTP listening at http://0.0.0.0:8080
2023-01-05T23:22:02.199Z [LOG] Debugger enabled at http://0.0.0.0:8080/_debug/
2023-01-05T23:22:02.199Z [LOG] Server started.
```


## Making changes
In order to make sure that all packages are built correctly and that Jest understands cross-package references you should run the included build watcher commands during local development

```shell
deepkit-framework » npm run tsc-watch
deepkit-framework » npm run tsc-watch:esm
```

## Important
- When one of the package.json files is modified (adding, removing or updating dependencies) you will need to re-run the `npm bootstrap` and `npm link` commands from above
- Never install NPM dependencies directly inside of any of the packages in the `packages/*` directory.

**Comment by colorcube** [comment]
<pre>
# Development

## Prerequisites
Deepkit uses NPM and Lerna to manage this monorepo. Local package linking is managed through the [npm-local-development](https://www.npmjs.com/package/npm-local-development) CLI.

Make sure `libpq5` and `libpq-dev` are installed, and `python` refers Python2.

Node v18 is needed.

## Getting Started

```shell
git clone https://github.com/deepkit/deepkit-framework.git
cd deepkit-framework
npm install
npm run bootstrap
npm run link
npm run install-compiler
```

When installation is finished you can build the packages:

```shell
deepkit-framework » npm run build
```

This could take several minutes. 
You should see the build messages and a _success_ summary in the end:

```shell
lerna success run Ran npm script 'build' in 36 packages in 83.9s:
lerna success - @deepkit/angular-universal
lerna success - @deepkit/api-console-api
...
```

If everything went fine you can try out the example app:

```shell
deepkit-framework » cd packages/example-app
deepkit-framework/packages/example-app » npm run app
```

That should give you a _usage_ message of the app. 

To start the app server:

```shell
deepkit-framework/packages/example-app » npm run start
```

```shell
...
2023-01-05T23:22:02.199Z [LOG] HTTP listening at http://0.0.0.0:8080
2023-01-05T23:22:02.199Z [LOG] Debugger enabled at http://0.0.0.0:8080/_debug/
2023-01-05T23:22:02.199Z [LOG] Server started.
```


## Making changes
In order to make sure that all packages are built correctly and that Jest understands cross-package references you should run the included build watcher commands during local development

```shell
deepkit-framework » npm run tsc-watch
deepkit-framework » npm run tsc-watch:esm
```

## Important
- When one of the package.json files is modified (adding, removing or updating dependencies) you will need to re-run the `npm bootstrap` and `npm link` commands from above
- Never install NPM dependencies directly inside of any of the packages in the `packages/*` directory.
</pre>

**Comment by colorcube** [comment]
merged in #405
